File: src/Facturatix.Contracts/Recipes/VariableCatalog.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Facturatix.Contracts.Recipes;

/// <summary>
/// Where a recipe variable's value comes from, and what the platform does with it.
/// <para>
/// <c>variableSource</c> in the JSON Schema bounds the NAMESPACE and nothing else. Measured against
/// the reference validator, <c>user.fiscal.rfk</c> and <c>user.nombre</c> are both valid documents:
/// they publish, reach a worker, resolve to nothing, and fail the ticket with
/// <c>variable_missing</c> — for a typo that three validation layers looked at and accepted. The
/// parts of the source that are genuinely closed therefore belong in the contract next to the
/// namespaces, and this type is those parts.
/// </para>
/// <para>
/// The two halves are closed differently. <see cref="UserFields"/> is a closed SET: its leaves are
/// columns the API owns, so a source outside the set cannot resolve.
/// <see cref="TicketValueKinds"/> closes the VOCABULARY and not the membership: a recipe may ask
/// the model for any field a portal needs, but only these leaf names carry a normalizer. That
/// distinction is load-bearing — the leaf of <c>ticket.extracted.&lt;leaf&gt;</c> is both the JSON
/// key the vision model is asked to return and the selector of the normalizer applied to the
/// answer, so <c>total</c> becomes <c>1234.50</c> while <c>total_venta</c> is passed through as
/// <c>$ 1,234.50</c>.
/// </para>
/// <para>
/// This type is a MIRROR of <c>schemas/variable-catalog.v1.json</c>, which ships in the same
/// package; the TypeScript twin mirrors the same file. A consumer proving it covers every entry
/// reads the artefact through <see cref="RecipeContractResources.ReadVariableCatalog"/> rather than
/// trusting this list, which is what makes such a test a gate instead of a tautology.
/// </para>
/// </summary>
public static class VariableCatalog
{
    /// <summary>File name of the artefact this type mirrors.</summary>
    public const string FileName = "variable-catalog.v1.json";

    /// <summary>Version of the catalogue mirrored here.</summary>
    public const int Version = 1;

    /// <summary>Whether a recipe may bind a stored user value.</summary>
    /// <remarks>
    /// <c>stored_only</c> is not an oversight to be fixed by adding the field to the resolver. A
    /// variable is only ever rendered into the <c>value</c> of a <c>fill</c>/<c>select</c> or the
    /// <c>url</c> of a <c>goto</c>, and the boolean fiscal flags have no text form a Mexican portal
    /// accepts; <c>check</c> and <c>uncheck</c> take no value, and every condition type reads the
    /// page rather than a variable. Making them bindable needs a contract addition — a condition
    /// that reads a variable — not a catalogue edit. They are listed so an author auditing their
    /// coverage can see that they exist and why they are not on offer.
    /// </remarks>
    public static class Support
    {
        public const string Bindable = "bindable";
        public const string StoredOnly = "stored_only";
    }

    /// <summary>Whether the platform can promise the field has a value.</summary>
    /// <remarks>
    /// <c>always</c> — every write path validates it as non-empty. <c>optional</c> — the platform
    /// accepts a profile without it, which covers the nullable columns and also <c>user.email</c>
    /// and <c>user.name</c>: both are <c>NOT NULL</c> and both arrive through a helper whose
    /// documented contract is "a value that fits, possibly empty".
    /// </remarks>
    public static class Availability
    {
        public const string Always = "always";
        public const string Optional = "optional";
    }

    /// <summary>What the platform does to a value the vision model read off the ticket.</summary>
    public static class ValueKinds
    {
        public const string Text = "text";
        public const string TaxId = "tax_id";
        public const string PostalCode = "postal_code";
        public const string Amount = "amount";
        public const string Date = "date";
    }

    /// <summary>
    /// Every value the platform stores about the account holder, in the order the artefact lists
    /// them. The order is part of the mirror: parity is compared positionally.
    /// </summary>
    public static readonly IReadOnlyList<UserFieldDefinition> UserFields = new[]
    {
        new UserFieldDefinition("user.email", Support.Bindable, Availability.Optional),
        new UserFieldDefinition("user.name", Support.Bindable, Availability.Optional),
        new UserFieldDefinition("user.fiscal.rfc", Support.Bindable, Availability.Always),
        new UserFieldDefinition("user.fiscal.legal_name", Support.Bindable, Availability.Always),
        new UserFieldDefinition("user.fiscal.tax_regime", Support.Bindable, Availability.Always),
        new UserFieldDefinition("user.fiscal.email", Support.Bindable, Availability.Optional),
        new UserFieldDefinition("user.fiscal.postal_code", Support.Bindable, Availability.Always),
        new UserFieldDefinition("user.fiscal.street", Support.Bindable, Availability.Optional),
        new UserFieldDefinition(
            "user.fiscal.exterior_number", Support.Bindable, Availability.Optional),
        new UserFieldDefinition(
            "user.fiscal.interior_number", Support.Bindable, Availability.Optional),
        new UserFieldDefinition("user.fiscal.neighborhood", Support.Bindable, Availability.Optional),
        new UserFieldDefinition("user.fiscal.municipality", Support.Bindable, Availability.Optional),
        new UserFieldDefinition("user.fiscal.state", Support.Bindable, Availability.Optional),
        new UserFieldDefinition("user.fiscal.reference", Support.Bindable, Availability.Optional),
        new UserFieldDefinition("user.fiscal.country", Support.Bindable, Availability.Always),
        new UserFieldDefinition("user.fiscal.is_foreign", Support.StoredOnly, Availability.Always),
        new UserFieldDefinition(
            "user.fiscal.is_politically_exposed", Support.StoredOnly, Availability.Always),
        new UserFieldDefinition(
            "user.fiscal.declares_ieps", Support.StoredOnly, Availability.Always),
    };

    /// <summary>
    /// Leaf names that carry a normalizer, in artefact order. A leaf outside these is normalized as
    /// <see cref="TicketDefaultValueKind"/>.
    /// </summary>
    public static readonly IReadOnlyList<TicketValueKindDefinition> TicketValueKinds = new[]
    {
        new TicketValueKindDefinition(ValueKinds.TaxId, new[] { "rfc", "tax_id", "taxid" }),
        new TicketValueKindDefinition(
            ValueKinds.PostalCode, new[] { "postal_code", "codigo_postal", "cp", "zip" }),
        new TicketValueKindDefinition(
            ValueKinds.Amount,
            new[] { "total", "amount", "importe", "monto", "subtotal", "iva", "impuesto" }),
        new TicketValueKindDefinition(
            ValueKinds.Date,
            new[] { "date", "fecha", "issue_date", "fecha_emision", "ticket_date" }),
    };

    /// <summary>
    /// The normalizer applied when a leaf matches no catalogued name: whitespace is collapsed and
    /// nothing else, because that is the only transformation that is safe on a value whose meaning
    /// is unknown.
    /// </summary>
    public const string TicketDefaultValueKind = ValueKinds.Text;

    private static readonly IReadOnlyDictionary<string, UserFieldDefinition> UserFieldBySource =
        UserFields.ToDictionary(field => field.Source, StringComparer.Ordinal);

    private static readonly IReadOnlyDictionary<string, string> TicketKindByLeaf =
        TicketValueKinds
            .SelectMany(entry => entry.Leaves.Select(leaf => (Leaf: leaf, entry.Kind)))
            .ToDictionary(pair => pair.Leaf, pair => pair.Kind, StringComparer.Ordinal);

    /// <summary>The sources a recipe variable may declare, in catalogue order.</summary>
    public static IReadOnlyList<string> BindableUserFieldSources { get; } =
        UserFields
            .Where(field => field.Support == Support.Bindable)
            .Select(field => field.Source)
            .ToList();

    /// <summary>
    /// The definition behind a user source, or <see langword="null"/> when the platform stores no
    /// such value. O(1).
    /// </summary>
    /// <param name="source">A variable's declared source.</param>
    public static UserFieldDefinition? FindUserField(string? source) =>
        source is not null && UserFieldBySource.TryGetValue(source, out var field) ? field : null;

    /// <summary>
    /// Whether a source names a user value a recipe may bind. Narrower than
    /// <see cref="RecipeSchemaV2.VariableSources"/>'s prefixes on purpose: the prefix answers
    /// "could this be a user field", this answers "is it one the executor will resolve". O(1).
    /// </summary>
    /// <param name="source">A variable's declared source.</param>
    public static bool IsBindableUserField(string? source) =>
        FindUserField(source)?.Support == Support.Bindable;

    /// <summary>
    /// What the platform will do to the value read for a ticket source. O(1).
    /// </summary>
    /// <param name="source">A <c>ticket.extracted.*</c> source, or its bare leaf.</param>
    /// <returns>
    /// The normalizer that will be applied, or <see cref="TicketDefaultValueKind"/> when the leaf
    /// carries none.
    /// </returns>
    public static string TicketValueKindOf(string? source)
    {
        if (source is null)
        {
            return TicketDefaultValueKind;
        }

        return TicketKindByLeaf.TryGetValue(LeafOfSource(source), out var kind)
            ? kind
            : TicketDefaultValueKind;
    }

    /// <summary>
    /// The last dot-separated segment of a canonical source.
    /// </summary>
    /// <param name="source">A variable's declared source, e.g. <c>ticket.extracted.total</c>.</param>
    /// <returns>The leaf, or the whole string when it carries no dot.</returns>
    public static string LeafOfSource(string source)
    {
        ArgumentNullException.ThrowIfNull(source);

        var lastDot = source.LastIndexOf('.');
        return lastDot >= 0 && lastDot < source.Length - 1 ? source[(lastDot + 1)..] : source;
    }
}

/// <summary>One value the platform stores about the account holder.</summary>
/// <param name="Source">The canonical <c>source</c> a recipe variable declares.</param>
/// <param name="Support">One of <see cref="VariableCatalog.Support"/>.</param>
/// <param name="Availability">One of <see cref="VariableCatalog.Availability"/>.</param>
public sealed record UserFieldDefinition(string Source, string Support, string Availability);

/// <summary>Leaf names that select one normalizer.</summary>
/// <param name="Kind">One of <see cref="VariableCatalog.ValueKinds"/>, never <c>text</c>.</param>
/// <param name="Leaves">Leaf names that map to <paramref name="Kind"/>, in artefact order.</param>
public sealed record TicketValueKindDefinition(string Kind, IReadOnlyList<string> Leaves);

File: src/ts/variable-catalog.ts
/**
 * Where a recipe variable's value comes from, and what the platform does with it.
 *
 * ## Why the schema's pattern is not enough
 *
 * `variableSource` in the JSON Schema bounds the NAMESPACE and nothing else:
 * `^(user\.fiscal\.[a-z][a-z0-9_]*|user\.[a-z][a-z0-9_]*|ticket\.extracted\.[a-z][a-z0-9_]*)$`.
 * Measured against the reference validator, `user.fiscal.rfk` and `user.nombre` are both **valid
 * documents**. They publish, reach a worker, resolve to nothing, and fail the ticket with
 * `variable_missing` — for a typo that three validation layers looked at and accepted.
 *
 * The parts of the source that are genuinely closed therefore belong in the contract next to the
 * namespaces, and this module is those parts. It answers, once, what the Modeler asks at authoring
 * time and the Generator asks at execution time:
 *
 * - which user values exist, and which of them may be bound at all
 * - what the platform will do to a value read off a ticket
 *
 * ## The two halves are closed differently, on purpose
 *
 * **{@link USER_FIELDS} is a closed set.** Its leaves are columns the API owns; a source outside
 * the set cannot resolve, so offering it or accepting it is a defect either way.
 *
 * **{@link TICKET_VALUE_KINDS} closes the vocabulary, not the membership.** A recipe may ask the
 * model for any field a portal needs — closing that would mean refusing a value the platform can
 * perfectly well extract. What is closed is which leaf names carry a normalizer, and that matters
 * more than it looks: the leaf of `ticket.extracted.<leaf>` is *both* the JSON key the vision model
 * is asked to return *and* the selector of the normalizer applied to the answer. `total` becomes
 * `1234.50`; `total_venta` is passed through as the model wrote it, `$ 1,234.50`, straight into the
 * portal's amount field. Two leaf names for the same concept are not synonyms here — one works and
 * one quietly does not.
 *
 * ## How the mirrors stay in step
 *
 * `schemas/variable-catalog.v1.json` is the artefact; this module and `VariableCatalog.cs` are both
 * mirrors of it. `scripts/verify-contract-parity.mjs` compares this module against the file entry
 * for entry and in order, and each consumer holds its own coverage test against the same artefact
 * — the Generator that it resolves every bindable field and normalizes every catalogued leaf, the
 * Modeler that it offers every bindable field and steers authors onto the catalogued leaves.
 *
 * @module
 */

// ── user fields ───────────────────────────────────────────────────────────────────────────────

/**
 * Whether a recipe may bind a stored user value.
 *
 * `bindable` — the Generator resolves it, so a variable may declare it as its source.
 *
 * `stored_only` — the platform holds the value but no recipe can act on it. This is not an
 * oversight to be fixed by adding it to the resolver: a variable is only ever rendered into the
 * `value` of a `fill`/`select` or the `url` of a `goto`, and the three boolean fiscal flags have no
 * text form a Mexican portal accepts. `check` and `uncheck` take no value, and the condition types
 * (`element_exists`, `element_absent`, `url_matches`) read the page rather than a variable, so
 * there is no verb through which a yes/no could act. Making these bindable needs a contract
 * addition — a condition that reads a variable — not a catalogue edit.
 *
 * They are catalogued anyway, and this is the point of the field: an author auditing which of their
 * profile data a recipe covers must be able to see that these exist and why they are not on offer.
 * A silent omission looks like a complete list.
 */
export type UserFieldSupport = 'bindable' | 'stored_only'

/**
 * Whether the platform can promise the field has a value.
 *
 * `always` — every write path validates it as non-empty, so a completed fiscal profile has it. A
 * required variable bound to one of these cannot fail resolution for want of a value.
 *
 * `optional` — the platform accepts a profile without it. That covers the nullable columns and also
 * two that are `NOT NULL` and still reachable as an empty string: `user.email` and `user.name` come
 * from the identity provider through a helper whose documented contract is "a value that fits,
 * possibly empty". A required variable bound to an `optional` field will fail some users' tickets,
 * which is sometimes exactly right — the portal refuses the invoice without it — and sometimes a
 * recipe defect. The Modeler says which it is at the moment of binding, so the choice is made
 * knowingly rather than discovered in the review queue.
 */
export type UserFieldAvailability = 'always' | 'optional'

/** One value the platform stores about the account holder. */
export interface UserFieldDefinition {
  /** The canonical `source` a recipe variable declares, e.g. `user.fiscal.rfc`. */
  readonly source: string
  readonly support: UserFieldSupport
  readonly availability: UserFieldAvailability
}

/**
 * Every value the platform stores about the account holder, in the order the artefact lists them.
 *
 * The order is part of the mirror: parity is compared positionally, so an entry moved in one half
 * and not the other is a failure rather than a silent difference in enumeration order.
 */
export const USER_FIELDS: readonly UserFieldDefinition[] = [
  { source: 'user.email', support: 'bindable', availability: 'optional' },
  { source: 'user.name', support: 'bindable', availability: 'optional' },
  { source: 'user.fiscal.rfc', support: 'bindable', availability: 'always' },
  { source: 'user.fiscal.legal_name', support: 'bindable', availability: 'always' },
  { source: 'user.fiscal.tax_regime', support: 'bindable', availability: 'always' },
  { source: 'user.fiscal.email', support: 'bindable', availability: 'optional' },
  { source: 'user.fiscal.postal_code', support: 'bindable', availability: 'always' },
  { source: 'user.fiscal.street', support: 'bindable', availability: 'optional' },
  { source: 'user.fiscal.exterior_number', support: 'bindable', availability: 'optional' },
  { source: 'user.fiscal.interior_number', support: 'bindable', availability: 'optional' },
  { source: 'user.fiscal.neighborhood', support: 'bindable', availability: 'optional' },
  { source: 'user.fiscal.municipality', support: 'bindable', availability: 'optional' },
  { source: 'user.fiscal.state', support: 'bindable', availability: 'optional' },
  { source: 'user.fiscal.reference', support: 'bindable', availability: 'optional' },
  { source: 'user.fiscal.country', support: 'bindable', availability: 'always' },
  { source: 'user.fiscal.is_foreign', support: 'stored_only', availability: 'always' },
  {
    source: 'user.fiscal.is_politically_exposed',
    support: 'stored_only',
    availability: 'always'
  },
  { source: 'user.fiscal.declares_ieps', support: 'stored_only', availability: 'always' }
] as const

/** The inventory keyed by source, for an O(1) lookup. */
const USER_FIELD_BY_SOURCE: ReadonlyMap<string, UserFieldDefinition> = new Map(
  USER_FIELDS.map((field) => [field.source, field])
)

/** The sources a recipe variable may declare, in catalogue order. */
export const BINDABLE_USER_FIELD_SOURCES: readonly string[] = USER_FIELDS.filter(
  (field) => field.support === 'bindable'
).map((field) => field.source)

/**
 * The definition behind a user source, or `null` when the platform stores no such value.
 *
 * Complexity: O(1).
 *
 * @param source - A variable's declared source.
 */
export function findUserField(source: string | null | undefined): UserFieldDefinition | null {
  if (source === null || source === undefined) return null
  return USER_FIELD_BY_SOURCE.get(source) ?? null
}

/**
 * Whether a source names a user value a recipe may bind.
 *
 * Narrower than the schema's prefix pattern on purpose: the pattern answers "could this be a user
 * field", this answers "is it one the Generator will resolve". A `user.*` source that fails here is
 * either a typo or a `stored_only` field, and both are worth saying before the recipe ships.
 *
 * Complexity: O(1).
 *
 * @param source - A variable's declared source.
 */
export function isBindableUserField(source: string | null | undefined): boolean {
  return findUserField(source)?.support === 'bindable'
}

// ── ticket fields ─────────────────────────────────────────────────────────────────────────────

/** What the platform does to a value the vision model read off the ticket. */
export type TicketValueKind = 'text' | 'tax_id' | 'postal_code' | 'amount' | 'date'

/** The leaf names that select one normalizer. */
export interface TicketValueKindDefinition {
  readonly kind: Exclude<TicketValueKind, 'text'>
  /** Leaf names that map to {@link kind}, in artefact order. */
  readonly leaves: readonly string[]
}

/**
 * The normalizer applied when a leaf matches no catalogued name.
 *
 * Whitespace is collapsed and nothing else. It is the only transformation that is safe on a value
 * whose meaning is unknown — and it is also why a mis-named amount reaches a portal with its
 * currency symbol still attached.
 */
export const TICKET_DEFAULT_VALUE_KIND: TicketValueKind = 'text'

/** Leaf names that carry a normalizer, in artefact order. */
export const TICKET_VALUE_KINDS: readonly TicketValueKindDefinition[] = [
  { kind: 'tax_id', leaves: ['rfc', 'tax_id', 'taxid'] },
  { kind: 'postal_code', leaves: ['postal_code', 'codigo_postal', 'cp', 'zip'] },
  {
    kind: 'amount',
    leaves: ['total', 'amount', 'importe', 'monto', 'subtotal', 'iva', 'impuesto']
  },
  { kind: 'date', leaves: ['date', 'fecha', 'issue_date', 'fecha_emision', 'ticket_date'] }
] as const

/** Leaf name to value kind, for an O(1) lookup. */
const TICKET_KIND_BY_LEAF: ReadonlyMap<string, TicketValueKind> = new Map(
  TICKET_VALUE_KINDS.flatMap((entry) =>
    entry.leaves.map((leaf) => [leaf, entry.kind as TicketValueKind] as const)
  )
)

/**
 * The last dot-separated segment of a canonical source.
 *
 * @param source - A variable's declared source, e.g. `ticket.extracted.total`.
 * @returns The leaf, or the whole string when it carries no dot.
 */
export function leafOfSource(source: string): string {
  const lastDot = source.lastIndexOf('.')
  return lastDot >= 0 && lastDot < source.length - 1 ? source.slice(lastDot + 1) : source
}

/**
 * What the platform will do to the value read for a ticket source.
 *
 * Complexity: O(1).
 *
 * @param source - A `ticket.extracted.*` source, or its bare leaf.
 * @returns The normalizer that will be applied, {@link TICKET_DEFAULT_VALUE_KIND} when the leaf
 * carries none.
 */
export function ticketValueKindOf(source: string | null | undefined): TicketValueKind {
  if (source === null || source === undefined) return TICKET_DEFAULT_VALUE_KIND
  return TICKET_KIND_BY_LEAF.get(leafOfSource(source)) ?? TICKET_DEFAULT_VALUE_KIND
}

// ── artefact identity ─────────────────────────────────────────────────────────────────────────

/** File name of the artefact both mirrors reproduce. */
export const VARIABLE_CATALOG_FILE_NAME = 'variable-catalog.v1.json'

/** Version of the catalogue this module mirrors. */
export const VARIABLE_CATALOG_VERSION = 1
